### PR TITLE
Vault secret manager

### PR DESCRIPTION
**Pull request by alvaroaleman**
Adds a vault secret manager, see the provided Readme for a detailled description of how it works. This PR only provides the backend part and no frontend yet. The first commit "Add a vaultclient" is in a dedicated PR: https://github.com/openshift/ci-tools/pull/1737

/cc @openshift/openshift-team-developer-productivity-test-platform 
Ref https://issues.redhat.com/browse/DPTP-1891

### CODE REVIEW

File: cmd/secret-collection-manager/main.go
Comment (stevekuznetsov) [inline cmd/secret-collection-manager/main.go:36]: Kind of weird that this has a default

File: cmd/secret-collection-manager/main.go
Comment (stevekuznetsov) [inline cmd/secret-collection-manager/main.go:50]: Might as well wrap with `interrupts.ListenAndServe`?

File: cmd/secret-collection-manager/main.go
Comment (stevekuznetsov) [inline cmd/secret-collection-manager/main.go:56]: What's the use of breaking this func out?

File: cmd/secret-collection-manager/main.go
Comment (stevekuznetsov) [inline cmd/secret-collection-manager/main.go:127]: We have health & readiness lib already from Prow, prefer that for consistency of code, ports, etc

File: cmd/secret-collection-manager/main.go
Comment (stevekuznetsov) [inline cmd/secret-collection-manager/main.go:229]: Can we unit test string manipulations like this block?

File: cmd/secret-collection-manager/main_test.go
Comment (stevekuznetsov) [inline cmd/secret-collection-manager/main_test.go:103]: The e2e framework has code to transparently stop, wait and stop (on timeout) processes that use the k8s ready/health endpoints (so it knows which ports to poke) - might be able to re-use?

File: pkg/vaultclient/vaultclient.go
Comment (stevekuznetsov) [inline pkg/vaultclient/vaultclient.go:165]: How are retries, etc handled? Inside of their client?

File: cmd/secret-collection-manager/main.go
Comment (alvaroaleman) [reply cmd/secret-collection-manager/main.go:50]: I would be less hesitant to use `interrupts` if https://github.com/kubernetes/test-infra/issues/17737 was fixed and it wasn't entirely based on package-level variables :/

File: cmd/secret-collection-manager/main.go
Comment (alvaroaleman) [reply cmd/secret-collection-manager/main.go:56]: That func is what the tests use

File: cmd/secret-collection-manager/main.go
Comment (alvaroaleman) [reply cmd/secret-collection-manager/main.go:127]: It doesn't really do anything different other than hardcoding a different port

File: pkg/vaultclient/vaultclient.go
Comment (alvaroaleman) [reply pkg/vaultclient/vaultclient.go]: Yeah, they use go-retriable underneath with a couple of custom settings, I thought it was best to just leave the defaults there

File: cmd/secret-collection-manager/main.go
Comment (petr-muller) [inline cmd/secret-collection-manager/main.go:159]: Curious, why 501?

File: cmd/secret-collection-manager/main.go
Comment (petr-muller) [inline cmd/secret-collection-manager/main.go:176]: typo: `Collection`

File: cmd/secret-collection-manager/main.go
Comment (petr-muller) [inline cmd/secret-collection-manager/main.go:249]: debugging leftover?

File: cmd/secret-collection-manager/main.go
Comment (petr-muller) [inline cmd/secret-collection-manager/main.go:313]: I'd consider `if HasPrefix(...) { groupNames = append(...) }` a bit more readable (getting rid of the negation and `continue`) but /shrug

File: cmd/secret-collection-manager/main.go
Comment (petr-muller) [inline cmd/secret-collection-manager/main.go:404]: :exploding_head:

**Comment by alvaroaleman** [comment]
@petr-muller Addressed all your feedback. Also renamed the thing to `vault secret collection manager`, because without the vault I can not remember the name

**Review by petr-muller — APPROVED** [review]
LGTM, holding if you want more reviews, it's a big thing

/hold

**Pull request merged**
